packages/core-data/src/types/menu-location.ts

```diff
@@ -0,0 +1,19 @@
+/**
+ * Internal dependencies
+ */
+import { Context } from './common';
+
+export interface MenuLocation< C extends Context > {
+	/**
+	 * The name of the menu location.
+	 */
+	name: string;
+	/**
+	 * The description of the menu location.
+	 */
+	description: string;
+	/**
+	 * The ID of the assigned menu.
+	 */
+	menu: number;
+}
```

Comment:
Even though `MenuLocation` doesn't current have any potential nevers, with interface merging it might. I think it'd be good to wrap all the types that plugins might extend to include contextual nevers with `OmitNevers`.

Ditto for `NavMenu`, etc.

Additionally, we should stick to the `Full` prefix for the interfaces these are for (or whatever appropriate prefix we can think of, I'm not sure `Full` exactly matches my expectation, but I'm not that interested in bike shedding over the perfect prefix).

Again, keeping in mind that these types will inevitably be extended by some plugin and we should strive to make it as easy as possible for plugin developers to publish consistent and accurate types.

Reply:
The `MenuLocation` export is free of any `never` properties already. If a plugin extends one of these types, I imagine it would look like this:

```ts
interface ExtendedMenuLocation< C extends Context > extends MenuLocation<C> {
	newField: ContextualField< string, 'edit', C >;
}
```

In that example, `ExtendedMenuLocation<'view'>['newField']` is of type never and thus, the plugin authors still need to wrap their types with `OmitNevers`.

Or did you mean extending the types more like below?

```ts
// In core-data
interface FullMenuLocation< C extends Context > {
	title: string;
}

type MenuLocation<C extends Context> = OmitNevers<FullMenuLocation<C>>

// In the extension
interface FullMenuLocation< C extends Context > {
	newField: ContextualField< string, 'edit', C >;
}
```

Reply:
Yes, I meant the second version. It's standard practice for extensible libraries to provide some method for doing so. Extensions of Jest's `expect` library use the namespace for example:

https://github.com/facebook/jest/blob/55caeb74cd8f91817f624feb8d54752e28f16767/packages/pretty-format/src/__tests__/setPrettyPrint.ts#L12

Reply:
Hm interface merging has a huge big upside: core-data selectors would return the extended types out of the box.

Reply:
Yes exactly.

Reply:
And with that in mind we should make sure that anything that _can_ be extended by a plugin is an `interface` rather than a `type`.

Reply:
We have a few threads about extending so I'll just keep responding here to avoid more fragmentation.

> Additionally, we should stick to the Full prefix for the interfaces these are for (or whatever appropriate prefix we can think of, I'm not sure Full exactly matches my expectation, but I'm not that interested in bike shedding over the perfect prefix).
> Change the "full" version of these types to be inside a namespace (global.wpCoreData) so that they can be extended through the global namespace.

@sarayourfriend I took some creative license with naming and this is how I imagine it could look like:

```ts
export namespace wpCoreData.BaseTypes {
	export interface Comment< C extends Context > {
		id: number;
		// ...
	}
}

export type Comment< C extends Context > = OmitNevers<
	wpCoreData.BaseTypes.Comment< C >
>;
```

In this scenario, I don't think we need the `Base` prefix anymore. The namespace _is_ the prefix. Every "base" type would live in that namespace, whether it has contextual fields or not. It would also have a generic `Context` as a part of the type signature. Specific types would be exported directly by the package, and the base types would be wrapped in a namespace to avoid confusion about twin exports like `Comment` and `FullComment`.

Reply:
That seems great to me.

Reply:
They'd just be a container for types so... hopefully web pack and ESM are completely oblivious to them (they'd be stripped out in the process of removing TypeScript stuff from the code before compilation).

Reply:
Oh no, I think you're right @dmsnell ! I played with that for about two hours now and didn't find a way to make `namespace merging`+`interface merging`+`inheritance`+`ES modules` work:

```ts
// attachment.ts
export namespace BaseTypes {
	export interface ExtensibleAttachment< C extends Context > {
		/**
		 * The date the post was published, in the site's timezone.
		 */
		date: string | null;
	}
}

export type Attachment< C extends Context > = OmitNevers<
	BaseTypes.ExtensibleAttachment< C >
>;

// index.ts
import { BaseTypes, Attachment } from './attachment';
import { Context } from './helpers';

namespace BaseTypes {
	export interface ExtensibleAttachment< C extends Context > {
		newField: number;
	}
}

const p: Attachment< any > = {};
// p.newField does not exist
```

I also played with `declare module` but also couldn't get it right.

Unless there's a trick that I'm not aware of, we might need to export the base types explicitly. I just gave them all an `Extensible` prefix to make the intention clear, e.g. `ExtensibleAttachment`. What do you think?

Reply:
I'm not sure what to make of that. I can play around with it later. In particular I'd like to see what happens when you try to extend the types outside the context of the package. I suspect that interface merging works a little differently in a consumer-side than on the side that is actually producing the emitted type definitions? That could be a totally wrong hunch though.

The worse case scenario is that yeah, we use that prefix. It's annoying because it's wordy but 99% of people won't ever know about it (only authors of this library and anyone extending these types) so it's really not _that_ bad. Hate that we're polluting the exported namespace that :disappointed: 

It'd be nice if we could create an intermediate package or something for them. Something like `@wordpress/core-data-types` that housed all the `Extensible` ones and then `core-data` just exported the "final" types after they're run through `OmitNevers` or whatever. And then we'd not publicly publish `@wordpress/core-data-types`, it'd just exist as something to `declare module "@wordpress/core-data-types" { ... }` for plugin authors.

It _seems_ roundabout but it would also be really nice to have the public type-level API of `core-data` not be complicated and muddied up by this distiction that, again, 99% of people shouldn't have to care about. Especially if you're doing something like this, it could be confusing which one to pick... `Extensible` sounds like a _very tempting_ prefix for someone to use irresponsibly or unwittingly dig themselves into a complicated type-hole:

![image](https://user-images.githubusercontent.com/24264157/155027259-efbe559c-3878-4c2f-a917-96d9aa401fae.png)

I'd be nice if we could find a way to dissuade users from ever mistakingly picking the `Extensible` one. I tried adding doc strings to the exported types I was using for testing in that screenshot but they do not show the JSDoc in VSCode when you're getting import auto-complete suggestions, so that's not a solution for this unfortunately.

Reply:
@sarayourfriend I agree, exporting the fully-qualified names is far from ideal. I have some good news, though – I think I cracked it! We can use a namespace for extending if we also use a separate module to host it:

```ts
// base-types.ts:
export namespace BaseTypes {}

// comment.ts:
declare module './base-types' {
	export namespace BaseTypes {
		export interface Comment< C extends Context > {
			id: number;
			// ...
		}
	}
}

export type Comment< C extends Context > = OmitNevers<
	BaseTypes.Comment< C >
>;

// index.ts:
export type { BaseTypes } from './base-types';

// my-plugin.ts:
import { Context } from '@wordpress/core-data';
declare module '@wordpress/core-data' {
	export namespace BaseTypes {
		export interface Comment< C extends Context > {
			numberOfViews: number;
		}
	}
}

import type { Comment } from '@wordpress/core-data';
const c = {} as Comment< 'view' >;

// c.numberOfViews is a number
// c.id is still present
```

It is more boilerplate than I hoped to have, but it isn't a deal-breaker to me. In fact, I'm quite happy with the powers it enables.

As for the build tools, `npm run build` and `npm run dev` both seem to work flawlessly. I'm not sure what else do we need to check to be certain nothing is broken?

cc @dmsnell @sarayourfriend

Reply:
did you test this out too? I'm still unable to extend these modules.

I'm concerned about the additional complexity introduced by the namespaces if it still means we won't be able to extend.

![Screen Shot 2022-02-22 at 12 35 16 PM](https://user-images.githubusercontent.com/5431237/155206729-726a9b0f-76ef-4a0c-98d3-c3f8058be69e.png)

probably I'm doing something wrong, but I can't figure out how to make this work.

I propose that we either figure out a working extensibility solution or leave the types as they were without the name spacing; this is bound to confuse others or our future selves.

Reply:
> did you test this out too? I'm still unable to extend these modules.

It works with the `"types"` line in the package.json. I ended up removing it for now as discussed with @sarayourfriend in https://github.com/WordPress/gutenberg/pull/38666#discussion_r811934106. If you restore it in your dev setup, it will work.

Reply:
I had this line in there when testing. I don't believe the rest of the typing would work as it does (see in the picture how it knows about `missing_image_sizes`) if that were gone.

Did you try this out and have type extensibility working?

Reply:
I did test that, here's a screenshot of a scratch file I put in the edit-post package:

![CleanShot 2022-02-22 at 22 02 51](https://user-images.githubusercontent.com/205419/155218962-d248c452-a6ac-49cd-b110-011d4dbbe01c.png)

Perhaps the culprit is a name of the namespace? I renamed it from `BaseTypes` to `WpBaseTypes` after writing the comment above, could that be it?

Reply:
> Perhaps the culprit is a name of the namespace? I renamed it from BaseTypes to WpBaseTypes after writing the comment above, could that be it?

that was it 👍

now. given we're working, is `WPBaseTypes` most-fitting here? to me that seems like it could conflate with more broad types. did you consider something more specific like `CoreEntityTypes`?

Reply:
`CoreEntityTypes` sounds better than WPBaseTypes, although it nudges that it's for the regular entity types. I just renamed it to `CoreBaseEntityTypes` – how does that sound?

Reply:
sounds fine. having `Entity` in there is nice

Reply:
not a big deal, but we still have `wp-base-types.ts`. could be fine since we're not interacting with that from outside the package but a little inconsistency

Reply:
`BaseEntityTypes` might also work. I don't care too much apart from the feeling that `WPBaseTypes` was a bit broad.

Reply:
👍  `BaseEntityTypes` it is, I adjusted the file name too